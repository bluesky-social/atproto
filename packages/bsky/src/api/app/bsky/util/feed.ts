import { TimeCidKeyset } from '../../../../db/pagination'
import { FeedRow } from '../../../../services/feed/types'

export enum FeedAlgorithm {
  ReverseChronological = 'reverse-chronological',
}

export class FeedKeyset extends TimeCidKeyset<FeedRow> {
  labelResult(result: FeedRow) {
    return { primary: result.sortAt, secondary: result.cid }
  }
}

// For users with sparse feeds, avoid scanning more than one week for a single page
export const getFeedDateThreshold = (from: string | undefined, days = 1) => {
  const timelineDateThreshold = from ? new Date(from) : new Date()
  timelineDateThreshold.setDate(timelineDateThreshold.getDate() - days)
  return timelineDateThreshold.toISOString()
}
