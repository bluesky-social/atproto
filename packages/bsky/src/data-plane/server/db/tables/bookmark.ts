export interface Bookmark {
  creator: string
  uri: string
  key: string
  indexedAt: string
}

export const tableName = 'bookmark'

export type PartialDB = { [tableName]: Bookmark }
