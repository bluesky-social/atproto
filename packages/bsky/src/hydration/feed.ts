import { DataPlaneClient } from '../data-plane/client'
import { Record as PostRecord } from '../lexicon/types/app/bsky/feed/post'
import { Record as LikeRecord } from '../lexicon/types/app/bsky/feed/like'
import { Record as RepostRecord } from '../lexicon/types/app/bsky/feed/repost'
import { Record as FeedGenRecord } from '../lexicon/types/app/bsky/feed/generator'
import { Record as ThreadgateRecord } from '../lexicon/types/app/bsky/feed/threadgate'
import { Record as PostgateRecord } from '../lexicon/types/app/bsky/feed/postgate'
import {
  HydrationMap,
  ItemRef,
  RecordInfo,
  parseRecord,
  parseString,
  split,
} from './util'
import { dedupeStrs } from '@atproto/common'
import { postUriToThreadgateUri, postUriToPostgateUri } from '../util/uris'

export type Post = RecordInfo<PostRecord> & {
  violatesThreadGate: boolean
  violatesEmbeddingRules: boolean
  hasThreadGate: boolean
  hasPostGate: boolean
}
export type Posts = HydrationMap<Post>

export type PostViewerState = {
  like?: string
  repost?: string
  threadMuted?: boolean
}

export type PostViewerStates = HydrationMap<PostViewerState>

export type PostAgg = {
  likes: number
  replies: number
  reposts: number
  quotes: number
}

export type PostAggs = HydrationMap<PostAgg>

export type Like = RecordInfo<LikeRecord>
export type Likes = HydrationMap<Like>

export type Repost = RecordInfo<RepostRecord>
export type Reposts = HydrationMap<Repost>

export type FeedGenAgg = {
  likes: number
}

export type FeedGenAggs = HydrationMap<FeedGenAgg>

export type FeedGen = RecordInfo<FeedGenRecord>
export type FeedGens = HydrationMap<FeedGen>

export type FeedGenViewerState = {
  like?: string
}

export type FeedGenViewerStates = HydrationMap<FeedGenViewerState>

export type Threadgate = RecordInfo<ThreadgateRecord>
export type Threadgates = HydrationMap<Threadgate>
export type Postgate = RecordInfo<PostgateRecord>
export type Postgates = HydrationMap<Postgate>

export type ThreadRef = ItemRef & { threadRoot: string }

// @NOTE the feed item types in the protos for author feeds and timelines
// technically have additional fields, not supported by the mock dataplane.
export type FeedItem = { post: ItemRef; repost?: ItemRef }

export class FeedHydrator {
  constructor(public dataplane: DataPlaneClient) {}

  async getPosts(
    uris: string[],
    includeTakedowns = false,
    given = new HydrationMap<Post>(),
  ): Promise<Posts> {
    const [have, need] = split(uris, (uri) => given.has(uri))
    const base = have.reduce(
      (acc, uri) => acc.set(uri, given.get(uri) ?? null),
      new HydrationMap<Post>(),
    )
    if (!need.length) return base
    const res = await this.dataplane.getPostRecords({ uris: need })
    return need.reduce((acc, uri, i) => {
      const record = parseRecord<PostRecord>(res.records[i], includeTakedowns)
      const violatesThreadGate = res.meta[i].violatesThreadGate
      const violatesEmbeddingRules = res.meta[i].violatesEmbeddingRules
      const hasThreadGate = res.meta[i].hasThreadGate
      const hasPostGate = res.meta[i].hasPostGate
      return acc.set(
        uri,
        record
          ? {
              ...record,
              violatesThreadGate,
              violatesEmbeddingRules,
              hasThreadGate,
              hasPostGate,
            }
          : null,
      )
    }, base)
  }

  async getPostViewerStates(
    refs: ThreadRef[],
    viewer: string,
  ): Promise<PostViewerStates> {
    if (!refs.length) return new HydrationMap<PostViewerState>()
    const threadRoots = refs.map((r) => r.threadRoot)
    const [likes, reposts, threadMutesMap] = await Promise.all([
      this.dataplane.getLikesByActorAndSubjects({
        actorDid: viewer,
        refs,
      }),
      this.dataplane.getRepostsByActorAndSubjects({
        actorDid: viewer,
        refs,
      }),
      this.getThreadMutes(threadRoots, viewer),
    ])
    return refs.reduce((acc, { uri, threadRoot }, i) => {
      return acc.set(uri, {
        like: parseString(likes.uris[i]),
        repost: parseString(reposts.uris[i]),
        threadMuted: threadMutesMap.get(threadRoot) ?? false,
      })
    }, new HydrationMap<PostViewerState>())
  }

  private async getThreadMutes(
    threadRoots: string[],
    viewer: string,
  ): Promise<Map<string, boolean>> {
    const deduped = dedupeStrs(threadRoots)
    const threadMutes = await this.dataplane.getThreadMutesOnSubjects({
      actorDid: viewer,
      threadRoots: deduped,
    })
    return deduped.reduce((acc, cur, i) => {
      return acc.set(cur, threadMutes.muted[i] ?? false)
    }, new Map<string, boolean>())
  }

  async getPostAggregates(refs: ItemRef[]): Promise<PostAggs> {
    if (!refs.length) return new HydrationMap<PostAgg>()
    const counts = await this.dataplane.getInteractionCounts({ refs })
    return refs.reduce((acc, { uri }, i) => {
      return acc.set(uri, {
        likes: counts.likes[i] ?? 0,
        reposts: counts.reposts[i] ?? 0,
        replies: counts.replies[i] ?? 0,
        quotes: counts.quotes[i] ?? 0,
      })
    }, new HydrationMap<PostAgg>())
  }

  async getFeedGens(
    uris: string[],
    includeTakedowns = false,
  ): Promise<FeedGens> {
    if (!uris.length) return new HydrationMap<FeedGen>()
    const res = await this.dataplane.getFeedGeneratorRecords({ uris })
    return uris.reduce((acc, uri, i) => {
      const record = parseRecord<FeedGenRecord>(
        res.records[i],
        includeTakedowns,
      )
      return acc.set(uri, record ?? null)
    }, new HydrationMap<FeedGen>())
  }

  async getFeedGenViewerStates(
    uris: string[],
    viewer: string,
  ): Promise<FeedGenViewerStates> {
    if (!uris.length) return new HydrationMap<FeedGenViewerState>()
    const likes = await this.dataplane.getLikesByActorAndSubjects({
      actorDid: viewer,
      refs: uris.map((uri) => ({ uri })),
    })
    return uris.reduce((acc, uri, i) => {
      return acc.set(uri, {
        like: parseString(likes.uris[i]),
      })
    }, new HydrationMap<FeedGenViewerState>())
  }

  async getFeedGenAggregates(refs: ItemRef[]): Promise<FeedGenAggs> {
    if (!refs.length) return new HydrationMap<FeedGenAgg>()
    const counts = await this.dataplane.getInteractionCounts({ refs })
    return refs.reduce((acc, { uri }, i) => {
      return acc.set(uri, {
        likes: counts.likes[i] ?? 0,
      })
    }, new HydrationMap<FeedGenAgg>())
  }

  async getThreadgatesForPosts(
    postUris: string[],
    includeTakedowns = false,
  ): Promise<Threadgates> {
    if (!postUris.length) return new HydrationMap<Threadgate>()
    const uris = postUris.map(postUriToThreadgateUri)
    return this.getThreadgateRecords(uris, includeTakedowns)
  }

  async getThreadgateRecords(
    uris: string[],
    includeTakedowns = false,
  ): Promise<Threadgates> {
    const res = await this.dataplane.getThreadGateRecords({ uris })
    return uris.reduce((acc, uri, i) => {
      const record = parseRecord<ThreadgateRecord>(
        res.records[i],
        includeTakedowns,
      )
      return acc.set(uri, record ?? null)
    }, new HydrationMap<Threadgate>())
  }

  async getPostgatesForPosts(
    postUris: string[],
    includeTakedowns = false,
  ): Promise<Postgates> {
    if (!postUris.length) return new HydrationMap<Postgate>()
    const uris = postUris.map(postUriToPostgateUri)
    return this.getPostgateRecords(uris, includeTakedowns)
  }

  async getPostgateRecords(
    uris: string[],
    includeTakedowns = false,
  ): Promise<Postgates> {
    const res = await this.dataplane.getPostgateRecords({ uris })
    return uris.reduce((acc, uri, i) => {
      const record = parseRecord<PostgateRecord>(
        res.records[i],
        includeTakedowns,
      )
      return acc.set(uri, record ?? null)
    }, new HydrationMap<Postgate>())
  }

  async getLikes(uris: string[], includeTakedowns = false): Promise<Likes> {
    if (!uris.length) return new HydrationMap<Like>()
    const res = await this.dataplane.getLikeRecords({ uris })
    return uris.reduce((acc, uri, i) => {
      const record = parseRecord<LikeRecord>(res.records[i], includeTakedowns)
      return acc.set(uri, record ?? null)
    }, new HydrationMap<Like>())
  }

  async getReposts(uris: string[], includeTakedowns = false): Promise<Reposts> {
    if (!uris.length) return new HydrationMap<Repost>()
    const res = await this.dataplane.getRepostRecords({ uris })
    return uris.reduce((acc, uri, i) => {
      const record = parseRecord<RepostRecord>(res.records[i], includeTakedowns)
      return acc.set(uri, record ?? null)
    }, new HydrationMap<Repost>())
  }
}
