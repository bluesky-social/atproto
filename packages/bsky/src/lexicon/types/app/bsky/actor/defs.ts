/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { ValidationResult, BlobRef } from '@atproto/lexicon'
import { lexicons } from '../../../../lexicons'
import { isObj, hasProp } from '../../../../util'
import { CID } from 'multiformats/cid'
import * as ComAtprotoLabelDefs from '../../../com/atproto/label/defs'
import * as AppBskyGraphDefs from '../graph/defs'

export interface ProfileViewBasic {
  did: string
  handle: string
  displayName?: string
  avatar?: string
  viewer?: ViewerState
  labels?: ComAtprotoLabelDefs.Label[]
  [k: string]: unknown
}

export function isProfileViewBasic(v: unknown): v is ProfileViewBasic {
  return (
    isObj(v) &&
    hasProp(v, '$type') &&
    v.$type === 'app.bsky.actor.defs#profileViewBasic'
  )
}

export function validateProfileViewBasic(v: unknown): ValidationResult {
  return lexicons.validate('app.bsky.actor.defs#profileViewBasic', v)
}

export interface ProfileView {
  did: string
  handle: string
  displayName?: string
  description?: string
  avatar?: string
  indexedAt?: string
  viewer?: ViewerState
  labels?: ComAtprotoLabelDefs.Label[]
  [k: string]: unknown
}

export function isProfileView(v: unknown): v is ProfileView {
  return (
    isObj(v) &&
    hasProp(v, '$type') &&
    v.$type === 'app.bsky.actor.defs#profileView'
  )
}

export function validateProfileView(v: unknown): ValidationResult {
  return lexicons.validate('app.bsky.actor.defs#profileView', v)
}

export interface ProfileViewDetailed {
  did: string
  handle: string
  displayName?: string
  description?: string
  avatar?: string
  banner?: string
  followersCount?: number
  followsCount?: number
  postsCount?: number
  indexedAt?: string
  viewer?: ViewerState
  labels?: ComAtprotoLabelDefs.Label[]
  [k: string]: unknown
}

export function isProfileViewDetailed(v: unknown): v is ProfileViewDetailed {
  return (
    isObj(v) &&
    hasProp(v, '$type') &&
    v.$type === 'app.bsky.actor.defs#profileViewDetailed'
  )
}

export function validateProfileViewDetailed(v: unknown): ValidationResult {
  return lexicons.validate('app.bsky.actor.defs#profileViewDetailed', v)
}

export interface ViewerState {
  muted?: boolean
  mutedByList?: AppBskyGraphDefs.ListViewBasic
  blockedBy?: boolean
  blocking?: string
  following?: string
  followedBy?: string
  [k: string]: unknown
}

export function isViewerState(v: unknown): v is ViewerState {
  return (
    isObj(v) &&
    hasProp(v, '$type') &&
    v.$type === 'app.bsky.actor.defs#viewerState'
  )
}

export function validateViewerState(v: unknown): ValidationResult {
  return lexicons.validate('app.bsky.actor.defs#viewerState', v)
}

export type Preferences = (
  | AdultContentPref
  | ContentLabelPref
  | { $type: string; [k: string]: unknown }
)[]

export interface AdultContentPref {
  enabled: boolean
  [k: string]: unknown
}

export function isAdultContentPref(v: unknown): v is AdultContentPref {
  return (
    isObj(v) &&
    hasProp(v, '$type') &&
    v.$type === 'app.bsky.actor.defs#adultContentPref'
  )
}

export function validateAdultContentPref(v: unknown): ValidationResult {
  return lexicons.validate('app.bsky.actor.defs#adultContentPref', v)
}

export interface ContentLabelPref {
  label: string
  visibility: 'show' | 'warn' | 'hide' | (string & {})
  [k: string]: unknown
}

export function isContentLabelPref(v: unknown): v is ContentLabelPref {
  return (
    isObj(v) &&
    hasProp(v, '$type') &&
    v.$type === 'app.bsky.actor.defs#contentLabelPref'
  )
}

export function validateContentLabelPref(v: unknown): ValidationResult {
  return lexicons.validate('app.bsky.actor.defs#contentLabelPref', v)
}
