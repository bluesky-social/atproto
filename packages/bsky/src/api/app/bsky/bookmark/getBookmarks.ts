import { mapDefined } from '@atproto/common'
import { AppContext } from '../../../../context'
import { HydrateCtx, Hydrator } from '../../../../hydration/hydrator'
import { Server } from '../../../../lexicon'
import { QueryParams } from '../../../../lexicon/types/app/bsky/bookmark/getBookmarks'
import {
  HydrationFnInput,
  PresentationFnInput,
  SkeletonFnInput,
  createPipeline,
  noRules,
} from '../../../../pipeline'
import { Views } from '../../../../views'
import { resHeaders } from '../../../util'

export default function (server: Server, ctx: AppContext) {
  const getBookmarks = createPipeline(
    skeleton,
    hydration,
    noRules, // Blocks are included and handled on views. Mutes are included.
    presentation,
  )
  server.app.bsky.bookmark.getBookmarks({
    auth: ctx.authVerifier.standard,
    handler: async ({ params, auth, req }) => {
      const viewer = auth.credentials.iss
      const labelers = ctx.reqLabelers(req)
      const hydrateCtx = await ctx.hydrator.createContext({
        labelers,
        viewer,
      })

      const result = await getBookmarks(
        { ...params, hydrateCtx: hydrateCtx.copy({ viewer }) },
        ctx,
      )

      return {
        encoding: 'application/json',
        body: result,
        headers: resHeaders({ labelers: hydrateCtx.labelers }),
      }
    },
  })
}

const skeleton = async (input: SkeletonFnInput<Context, Params>) => {
  const { params, ctx } = input
  const actorDid = params.hydrateCtx.viewer
  const { uris, cursor } = await ctx.hydrator.dataplane.getBookmarkUris({
    actorDid: params.hydrateCtx.viewer,
    limit: params.limit,
    cursor: params.cursor,
  })
  return {
    actorDid,
    uris,
    cursor: cursor || undefined,
  }
}

const hydration = async (
  input: HydrationFnInput<Context, Params, SkeletonState>,
) => {
  const { ctx, params, skeleton } = input
  const { uris } = skeleton
  const state = await ctx.hydrator.hydratePosts(
    uris.map((uri) => ({ uri })),
    params.hydrateCtx,
  )
  return state
}

const presentation = (
  input: PresentationFnInput<Context, Params, SkeletonState>,
) => {
  const { ctx, hydration, skeleton } = input
  const { uris, cursor } = skeleton
  const bookmarks = mapDefined(uris, (uri) =>
    ctx.views.bookmark(uri, hydration),
  )
  return { bookmarks: bookmarks, cursor }
}

type Context = {
  hydrator: Hydrator
  views: Views
}

type Params = QueryParams & {
  hydrateCtx: HydrateCtx & { viewer: string }
}

type SkeletonState = {
  actorDid: string
  uris: string[]
  cursor?: string
}
