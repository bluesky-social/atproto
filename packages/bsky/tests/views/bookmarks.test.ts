import assert from 'node:assert'
import {
  $Typed,
  AppBskyBookmarkCreateBookmark,
  AppBskyBookmarkDeleteBookmark,
  AppBskyFeedDefs,
  AtpAgent,
} from '@atproto/api'
import { SeedClient, TestNetwork, basicSeed } from '@atproto/dev-env'
import { ids } from '../../src/lexicon/lexicons'
import { OutputSchema as GetBookmarksOutputSchema } from '../../src/lexicon/types/app/bsky/bookmark/getBookmarks'
import { PostView } from '../../src/lexicon/types/app/bsky/feed/defs'
import { forSnapshot, paginateAll } from '../_util'

type Database = TestNetwork['bsky']['db']

describe('appview bookmarks views', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let sc: SeedClient
  let db: Database

  // account dids, for convenience
  let alice: string
  let bob: string
  let carol: string
  let dan: string

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'bsky_views_bookmarks',
    })
    db = network.bsky.db
    agent = network.bsky.getClient()
    sc = network.getSeedClient()
    await basicSeed(sc)
    await network.processAll()

    alice = sc.dids.alice
    bob = sc.dids.bob
    carol = sc.dids.carol
    dan = sc.dids.dan
  })

  afterEach(async () => {
    jest.resetAllMocks()
    await clearPrivateData(db)
    await clearBookmarks(db)
  })

  afterAll(async () => {
    await network.close()
  })

  const get = async (actor: string, limit?: number, cursor?: string) =>
    agent.app.bsky.bookmark.getBookmarks(
      { limit, cursor },
      {
        headers: await network.serviceHeaders(
          actor,
          ids.AppBskyBookmarkGetBookmarks,
        ),
      },
    )

  const create = async (actor: string, uri: string) =>
    agent.app.bsky.bookmark.createBookmark(
      { uri },
      {
        headers: await network.serviceHeaders(
          actor,
          ids.AppBskyBookmarkCreateBookmark,
        ),
      },
    )

  const del = async (actor: string, uri: string) =>
    agent.app.bsky.bookmark.deleteBookmark(
      { uri },
      {
        headers: await network.serviceHeaders(
          actor,
          ids.AppBskyBookmarkDeleteBookmark,
        ),
      },
    )

  describe('creation', () => {
    it('creates bookmarks', async () => {
      await create(alice, sc.posts[alice][0].ref.uriStr)
      await create(alice, sc.posts[bob][0].ref.uriStr)
      await create(alice, sc.posts[carol][0].ref.uriStr)

      await create(bob, sc.posts[bob][0].ref.uriStr)
      await create(bob, sc.posts[carol][0].ref.uriStr)

      const { data: dataAlice } = await get(alice)
      expect(dataAlice.bookmarks).toHaveLength(3)

      const { data: dataBob } = await get(bob)
      expect(dataBob.bookmarks).toHaveLength(2)
    })

    it('fails on dupes', async () => {
      const uri = sc.posts[alice][0].ref.uriStr
      await create(alice, uri)
      await expect(create(alice, uri)).rejects.toThrow(
        AppBskyBookmarkCreateBookmark.DuplicatedError,
      )
    })

    it('fails on unsupported collections', async () => {
      const followUri = sc.follows[alice][bob].uriStr
      await expect(create(alice, followUri)).rejects.toThrow(
        AppBskyBookmarkCreateBookmark.UnsupportedCollectionError,
      )
    })
  })

  describe('deletion', () => {
    it('removes bookmarks', async () => {
      await create(alice, sc.posts[alice][0].ref.uriStr)
      await create(alice, sc.posts[bob][0].ref.uriStr)
      await create(alice, sc.posts[carol][0].ref.uriStr)

      const { data: dataBefore } = await get(alice)
      expect(dataBefore.bookmarks).toHaveLength(3)

      await del(alice, sc.posts[alice][0].ref.uriStr)
      await del(alice, sc.posts[carol][0].ref.uriStr)

      const { data: dataAfter } = await get(alice)
      expect(dataAfter.bookmarks).toHaveLength(1)
    })

    it('fails on not found', async () => {
      const uri = sc.posts[alice][0].ref.uriStr
      await expect(del(alice, uri)).rejects.toThrow(
        AppBskyBookmarkDeleteBookmark.NotFoundError,
      )
    })

    it('fails on unsupported collections', async () => {
      const followUri = sc.follows[alice][bob].uriStr
      await expect(del(alice, followUri)).rejects.toThrow(
        AppBskyBookmarkDeleteBookmark.UnsupportedCollectionError,
      )
    })
  })

  describe('listing', () => {
    it('gets empty bookmarks', async () => {
      const { data } = await get(alice)
      expect(data.bookmarks).toHaveLength(0)
    })

    it('paginates bookmarks', async () => {
      await create(alice, sc.posts[alice][0].ref.uriStr)
      await create(alice, sc.posts[alice][1].ref.uriStr)
      await create(alice, sc.posts[bob][0].ref.uriStr)
      await create(alice, sc.posts[bob][1].ref.uriStr)
      await create(alice, sc.posts[carol][0].ref.uriStr)
      await create(alice, sc.posts[dan][0].ref.uriStr)
      await create(alice, sc.posts[dan][1].ref.uriStr)

      const results = (out: GetBookmarksOutputSchema[]) =>
        out.flatMap((res) => res.bookmarks)

      const paginator = async (cursor?: string) => {
        const res = await get(alice, 2, cursor)
        return res.data
      }

      const fullRes = await get(alice)
      expect(fullRes.data.bookmarks.length).toBe(7)

      const paginatedRes = await paginateAll(paginator)
      paginatedRes.forEach((res) =>
        expect(res.bookmarks.length).toBeLessThanOrEqual(2),
      )

      const full = results([fullRes.data])
      assertPostViews(full)

      const paginated = results(paginatedRes)
      assertPostViews(paginated)

      const sort = (
        a: { item: $Typed<PostView> },
        b: { item: $Typed<PostView> },
      ) => (a.item.uri > b.item.uri ? 1 : -1)
      expect(paginated.sort(sort)).toEqual(full.sort(sort))
    })

    it('removes entries by blocked users, bidirectionally', async () => {
      await create(alice, sc.posts[alice][0].ref.uriStr)
      await create(alice, sc.posts[bob][0].ref.uriStr)
      await create(alice, sc.posts[carol][0].ref.uriStr)

      await create(bob, sc.posts[alice][0].ref.uriStr)
      await create(bob, sc.posts[carol][0].ref.uriStr)

      await sc.block(alice, bob)
      await network.processAll()

      const {
        data: { bookmarks: bookmarksA },
      } = await get(alice)
      expect(bookmarksA).toHaveLength(3)
      expect(bookmarksA[0].item.$type).toBe('app.bsky.feed.defs#postView')
      expect(bookmarksA[1].item.$type).toBe('app.bsky.feed.defs#blockedPost')
      expect(bookmarksA[2].item.$type).toBe('app.bsky.feed.defs#postView')
      expect(forSnapshot(bookmarksA)).toMatchSnapshot()

      const {
        data: { bookmarks: bookmarksB },
      } = await get(bob)
      expect(bookmarksB).toHaveLength(2)
      expect(bookmarksB[0].item.$type).toBe('app.bsky.feed.defs#postView')
      expect(bookmarksB[1].item.$type).toBe('app.bsky.feed.defs#blockedPost')
      expect(forSnapshot(bookmarksB)).toMatchSnapshot()
    })
  })
})

const clearPrivateData = async (db: Database) => {
  await db.db.deleteFrom('private_data').execute()
}

const clearBookmarks = async (db: Database) => {
  await db.db.deleteFrom('bookmark').execute()
}

function assertPostViews(
  bookmarks: GetBookmarksOutputSchema['bookmarks'],
): asserts bookmarks is { item: $Typed<AppBskyFeedDefs.PostView> }[] {
  bookmarks.forEach((b) => {
    assert(
      AppBskyFeedDefs.isPostView(b.item),
      `Expected bookmark to be a post view`,
    )
  })
}
