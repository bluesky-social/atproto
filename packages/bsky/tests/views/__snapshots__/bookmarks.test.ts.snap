// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appview bookmarks views listing removes entries by blocked users, bidirectionally 1`] = `
Array [
  Object {
    "item": Object {
      "$type": "app.bsky.feed.defs#postView",
      "author": Object {
        "associated": Object {
          "activitySubscription": Object {
            "allowSubscriptions": "followers",
          },
        },
        "did": "user(0)",
        "handle": "carol.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(2)",
          "following": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "embed": Object {
        "$type": "app.bsky.embed.recordWithMedia#view",
        "media": Object {
          "$type": "app.bsky.embed.images#view",
          "images": Array [
            Object {
              "alt": "../dev-env/assets/key-landscape-small.jpg",
              "fullsize": "https://bsky.public.url/img/feed_fullsize/plain/user(1)/cids(1)@jpeg",
              "thumb": "https://bsky.public.url/img/feed_thumbnail/plain/user(1)/cids(1)@jpeg",
            },
            Object {
              "alt": "../dev-env/assets/key-alt.jpg",
              "fullsize": "https://bsky.public.url/img/feed_fullsize/plain/user(1)/cids(2)@jpeg",
              "thumb": "https://bsky.public.url/img/feed_thumbnail/plain/user(1)/cids(2)@jpeg",
            },
          ],
        },
        "record": Object {
          "$type": "app.bsky.embed.record#view",
          "record": Object {
            "$type": "app.bsky.embed.record#viewBlocked",
            "author": Object {
              "did": "user(2)",
              "viewer": Object {
                "blockedBy": false,
                "blocking": "record(4)",
              },
            },
            "blocked": true,
            "uri": "record(3)",
          },
        },
      },
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "likeCount": 2,
      "quoteCount": 1,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "embed": Object {
          "$type": "app.bsky.embed.recordWithMedia",
          "media": Object {
            "$type": "app.bsky.embed.images",
            "images": Array [
              Object {
                "alt": "../dev-env/assets/key-landscape-small.jpg",
                "image": Object {
                  "$type": "blob",
                  "mimeType": "image/jpeg",
                  "ref": Object {
                    "$link": "cids(1)",
                  },
                  "size": 4114,
                },
              },
              Object {
                "alt": "../dev-env/assets/key-alt.jpg",
                "image": Object {
                  "$type": "blob",
                  "mimeType": "image/jpeg",
                  "ref": Object {
                    "$link": "cids(2)",
                  },
                  "size": 12736,
                },
              },
            ],
          },
          "record": Object {
            "record": Object {
              "cid": "cids(3)",
              "uri": "record(3)",
            },
          },
        },
        "text": "hi im carol",
      },
      "replyCount": 0,
      "repostCount": 0,
      "uri": "record(0)",
      "viewer": Object {
        "embeddingDisabled": false,
        "like": "record(5)",
        "threadMuted": false,
      },
    },
  },
  Object {
    "item": Object {
      "$type": "app.bsky.feed.defs#blockedPost",
      "author": Object {
        "did": "user(2)",
        "viewer": Object {
          "blockedBy": false,
          "blocking": "record(4)",
        },
      },
      "blocked": true,
      "uri": "record(3)",
    },
  },
  Object {
    "item": Object {
      "$type": "app.bsky.feed.defs#postView",
      "author": Object {
        "associated": Object {
          "activitySubscription": Object {
            "allowSubscriptions": "followers",
          },
        },
        "avatar": "https://bsky.public.url/img/avatar/plain/user(4)/cids(5)@jpeg",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "did": "user(3)",
        "displayName": "ali",
        "handle": "alice.test",
        "labels": Array [
          Object {
            "cid": "cids(6)",
            "cts": "1970-01-01T00:00:00.000Z",
            "src": "user(3)",
            "uri": "record(7)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(6)",
            "cts": "1970-01-01T00:00:00.000Z",
            "src": "user(3)",
            "uri": "record(7)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "muted": false,
        },
      },
      "cid": "cids(4)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [
        Object {
          "cid": "cids(4)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(3)",
          "uri": "record(6)",
          "val": "self-label",
        },
      ],
      "likeCount": 0,
      "quoteCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "labels": Object {
          "$type": "com.atproto.label.defs#selfLabels",
          "values": Array [
            Object {
              "val": "self-label",
            },
          ],
        },
        "text": "hey there",
      },
      "replyCount": 0,
      "repostCount": 0,
      "uri": "record(6)",
      "viewer": Object {
        "embeddingDisabled": false,
        "threadMuted": false,
      },
    },
  },
]
`;

exports[`appview bookmarks views listing removes entries by blocked users, bidirectionally 2`] = `
Array [
  Object {
    "item": Object {
      "$type": "app.bsky.feed.defs#postView",
      "author": Object {
        "associated": Object {
          "activitySubscription": Object {
            "allowSubscriptions": "followers",
          },
        },
        "did": "user(0)",
        "handle": "carol.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "following": "record(1)",
          "muted": false,
        },
      },
      "cid": "cids(0)",
      "embed": Object {
        "$type": "app.bsky.embed.recordWithMedia#view",
        "media": Object {
          "$type": "app.bsky.embed.images#view",
          "images": Array [
            Object {
              "alt": "../dev-env/assets/key-landscape-small.jpg",
              "fullsize": "https://bsky.public.url/img/feed_fullsize/plain/user(1)/cids(1)@jpeg",
              "thumb": "https://bsky.public.url/img/feed_thumbnail/plain/user(1)/cids(1)@jpeg",
            },
            Object {
              "alt": "../dev-env/assets/key-alt.jpg",
              "fullsize": "https://bsky.public.url/img/feed_fullsize/plain/user(1)/cids(2)@jpeg",
              "thumb": "https://bsky.public.url/img/feed_thumbnail/plain/user(1)/cids(2)@jpeg",
            },
          ],
        },
        "record": Object {
          "record": Object {
            "$type": "app.bsky.embed.record#viewRecord",
            "author": Object {
              "associated": Object {
                "activitySubscription": Object {
                  "allowSubscriptions": "followers",
                },
              },
              "avatar": "https://bsky.public.url/img/avatar/plain/user(3)/cids(4)@jpeg",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "did": "user(2)",
              "displayName": "bobby",
              "handle": "bob.test",
              "labels": Array [],
              "viewer": Object {
                "blockedBy": false,
                "muted": false,
              },
            },
            "cid": "cids(3)",
            "embeds": Array [],
            "indexedAt": "1970-01-01T00:00:00.000Z",
            "labels": Array [],
            "likeCount": 0,
            "quoteCount": 1,
            "replyCount": 0,
            "repostCount": 0,
            "uri": "record(2)",
            "value": Object {
              "$type": "app.bsky.feed.post",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "langs": Array [
                "en-US",
                "i-klingon",
              ],
              "text": "bob back at it again!",
            },
          },
        },
      },
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "likeCount": 2,
      "quoteCount": 1,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "embed": Object {
          "$type": "app.bsky.embed.recordWithMedia",
          "media": Object {
            "$type": "app.bsky.embed.images",
            "images": Array [
              Object {
                "alt": "../dev-env/assets/key-landscape-small.jpg",
                "image": Object {
                  "$type": "blob",
                  "mimeType": "image/jpeg",
                  "ref": Object {
                    "$link": "cids(1)",
                  },
                  "size": 4114,
                },
              },
              Object {
                "alt": "../dev-env/assets/key-alt.jpg",
                "image": Object {
                  "$type": "blob",
                  "mimeType": "image/jpeg",
                  "ref": Object {
                    "$link": "cids(2)",
                  },
                  "size": 12736,
                },
              },
            ],
          },
          "record": Object {
            "record": Object {
              "cid": "cids(3)",
              "uri": "record(2)",
            },
          },
        },
        "text": "hi im carol",
      },
      "replyCount": 0,
      "repostCount": 0,
      "uri": "record(0)",
      "viewer": Object {
        "embeddingDisabled": false,
        "like": "record(3)",
        "threadMuted": false,
      },
    },
  },
  Object {
    "item": Object {
      "$type": "app.bsky.feed.defs#blockedPost",
      "author": Object {
        "did": "user(4)",
        "viewer": Object {
          "blockedBy": true,
        },
      },
      "blocked": true,
      "uri": "record(4)",
    },
  },
]
`;
